synapse/federation/federation_client.py

```diff
@@ -358,11 +358,33 @@ async def get_pdu(
             The requested PDU, or None if we were unable to find it.
         """
 
+        logger.debug(
+            "get_pdu: event_id=%s from destinations=%s", event_id, destinations
+        )
+
         # TODO: Rate limit the number of times we try and get the same event.
 
-        ev = self._get_pdu_cache.get(event_id)
-        if ev:
-            return ev
+        event_copy = None
+        event_from_cache = self._get_pdu_cache.get(event_id)
+        if event_from_cache:
+            logger.debug("get_pdu: found event_from_cache=%s", event_from_cache)
+            assert not event_from_cache.internal_metadata.outlier, (
+                "Event from cache unexpectedly an `outlier` when it should be pristine and untouched without metadata set. "
+                "We are probably not be returning a copy of the event because downstream callers are modifying the event reference we have in the cache."
```

Comment:
> We are probably not be returning a copy

except we know we *are* returning a copy, because the code to do so is ten lines below. I'm convinced this sort of speculation is terribly helpful - all we know is that somewhere, some assumption has been violated.